PollyFill.cs

```diff
@@ -0,0 +1,115 @@
+﻿namespace GrayBMP;
+class PolyFill {
+   public void AddLine (int x0, int y0, int x1, int y1) {
+      var pt1 = (x0, y0); var pt2 = (x1, y1);
+      if (pt1 == pt2) return;
+      // Adjust segment ends along the sweep direction (at a lower y and then lower x).
+      if (Compare (pt1, pt2) > 0) (pt1, pt2) = (pt2, pt1);
+      Add (pt1); Add (pt2);
+
+      void Add ((int x, int y) pt) {
+         if (pt.x < xLeft) xLeft = pt.x;
+         if (pt.x > xRight) xRight = pt.x;
+         mPts.Add (pt);
+      }
+   }
+
+   public void Fill (GrayBMP bmp, int color) => Fill (mPts.ToArray (), bmp, color);
+
+   // Fill polygon using sweepline
+   void Fill ((int x, int y)[] pts, GrayBMP bmp, int color) {
+      // Build event queue by adding enter and exit events for every 'edge'.
+      // We use an array here as we never add any new event during the scan.
+      int[] events = Enumerable.Range (0, pts.Length).ToArray ();
+      // Sort events list to event queue.
+      Array.Sort (events, CompareEvent);
+
+      // Scan range.
+      int yStart = pts[events[0]].y, yEnd = -1;
+      // The active-edge-list
+      List<int> ael = new ();
+      // The intersection points at a given 'y'.
+      List<double> xi = new ();
+      // Process events from event queue.
+      for (int i = 1; i < events.Length; i++) {
+         int id = events[i];
+         // record the event point
+         yEnd = pts[id].y;
+         bool enter = id % 2 == 0; // Enter event?
+         if (enter) {
+            // Add an edge to the active edge list.
+            ael.Add (id);

```

Comment:
Your code works fine, but is slightly confusing to a novice because:
- id is the start point of a line if it is even
- id is the end point of a line if it is odd
You add only the start point of a line into the ael, and you have to subsequently use pts[edge + 1] to get the endpoint. It's not immediately clear from the way you build the ael that you are adding events for both the start and end poins (and the id % 2 == 0 to figure out which type is slightly kludgy).

This is a really minor criticism. But would the code be a bit cleaner to read (without being any less efficient) by having a Line struct and maintaining the ael as a list of such **Line** objects?

Reply:
Added **Event** struct to capture enter/exit events. **Event.Enter** contains the event type. This change made it slightly slower (1ms per Fill). But I am not complaining :-).